File: python/Zigzag-Conversion/Zigzag-Conversion.py
#! /usr/bin/env python
#! -*- coding=utf-8 -*-
# Date: 8/12/19
# Author: Bryce
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        n = len(s)
        res = ""
        if numRows == 1:
            return s
        m = 2 * numRows -2       
        for i in range (numRows):
            if (i == 0 or i == numRows-1):
                for j in range (i,n,m):
                    res += s[j]
            else:
                j = i
                k = 2*numRows -i -1 -1
                while (j < n or k < n ):
                    if j < n:
                        res += s[j]
                    if k < n:
                        res += s[k]
                    j += m
                    k += m
        return res

File: cpp/Zigzag-Conversion/Zigzag-Conversion.cpp
/// Source : https://leetcode-cn.com/problems/zigzag-conversion/submissions/
/// Author : bryce
/// Time   : 2019-10-26
class Solution {
public:
    string convert(string s, int numRows) {
        string res;
        int n = s.length();
        int m = 2*numRows -2;
        if (numRows == 1) return s;
        for(int i = 0; i < numRows; i++ )
        {
            if(i == 0 || i == numRows -1)
            {
                for (int j = i; j <n; j+=m)
                {
                    res +=s[j];
                }
            }
            else
            {
                for (int j = i, k = 2 * numRows - i - 1 - 1; j < n || k < n; j+=m, k +=m)
                {
                    if (j <n)
                        res += s[j];
                    if (k <n)
                        res += s[k];
                }
            }
        }
        
        return res;
    }
};
